modules/com_vtiger_workflow/tasks/CBWatermark.php

```diff
@@ -0,0 +1,311 @@
+				'modules' => '',
+				'massaction' => true,
+				'description' => 'HELP_WaterMark_Size',
+			),
+			'WaterMark_Position' => array(
+				'type' => '7',
+				'values' => 'integer',
+				'modules' => '',
+				'massaction' => true,
+				'description' => 'HELP_WaterMark_Position',
+			),
+			'Expression_Type' => array(
+				'type' => '1',
+				'values' => 'varchar',
+				'modules' => '',
+				'massaction' => true,
+				'description' => 'HELP_Expression_Type',
+			),
+		);
+	}
+
+	public function doTask(&$entity) {
+		global $current_user, $logbg, $from_wf, $currentModule, $root_directory;
+		global $log;
+		$logbg->debug('> CBWatermark');
+
+		// setting workflow variables
+		$wmImageValue = null;
+		$imagefieldName = null;
+		$wmsize = null;
+		$wmPosition = null;
+		$exptype = null;
+		if (empty($entity->WorkflowContext['WaterMark_Image_value'])) {
+			$wmImageValue = $this->wmImageValue;
+		} else {
+			$wmImageValue = $entity->WorkflowContext['WaterMark_Image_value'];
+		}
+
+		if (empty($entity->WorkflowContext['Image_Fieled_Name'])) {
+			$imagefieldName = $this->imagefieldName;
+		} else {
+			$imagefieldName = $entity->WorkflowContext['Image_Fieled_Name'];
+		}
+
+		if (empty($entity->WorkflowContext['WaterMark_Size'])) {
+			$wmsize = $this->wmSize;
+		} else {
+			$wmsize = $entity->WorkflowContext['WaterMark_Size'];
+		}
+
+		if (empty($entity->WorkflowContext['WaterMark_Position'])) {
+			$wmPosition = $this->wmPosition;
+		} else {
+			$wmPosition = $entity->WorkflowContext['WaterMark_Position'];
+		}
+
+		if (empty($entity->WorkflowContext['Expression_Type'])) {
+			$exptype = $this->exptype;
+		} else {
+			$exptype = $entity->WorkflowContext['Expression_Type'];
+		}
+
+		$currentRecordId = explode('x', $entity->id)[1];
+		$from_wf = true;
+		$hold_ajxaction = isset($_REQUEST['ajxaction']) ? $_REQUEST['ajxaction'] : '';
+		$_REQUEST['ajxaction'] = 'Workflow';
+		if (!empty($wmImageValue)) {
+			$watermark = '';
+			if ($exptype == 'rawtext') {
+				$watermark = $wmImageValue;
+			}
+			// elseif ($exptype == 'fieldname') {
+			// 	$util = new VTWorkflowUtils();
+			// 	$adminUser = $util->adminUser();
+			// 	$entityCache = new VTEntityCache($adminUser);
+			// 	$fn = new VTSimpleTemplate($wmImageValue);
+			// 	$watermark = $fn->render($entityCache, $entity->getId(), [], $entity->WorkflowContext);
+			// } else {
+			// 	$parser = new VTExpressionParser(new VTExpressionSpaceFilter(new VTExpressionTokenizer($wmImageValue)));
+			// 	$expression = $parser->expression();
+			// 	$exprEvaluater = new VTFieldExpressionEvaluater($expression);
+			// 	$watermark = $exprEvaluater->evaluate($entity);
+			// }
+			$waterMarkUrl = $watermark;
+			$waterMarkArr = explode('/', $waterMarkUrl);
+			$waterMarkDotArr = explode('.', $waterMarkUrl);
+			$waterMarkName = $waterMarkArr[count($waterMarkArr) - 1];
+			$waterMarkType = $waterMarkDotArr[count($waterMarkDotArr) - 1];
+			$waterMarkSavedPath = $root_directory . 'storage/';
```

Comment:
this hard-coded `storage` should be:  `GlobalVariable::getVariable('Application_Storage_Directory', 'storage/')`